### PR TITLE
improve run CR's create validation

### PR DESCRIPTION
**Pull request by shirmoran**
This improvement adds the following validations:

**ConfigMap field:**
- checks that "configMapName" field is not empty.
- checks that the configMap exists.
- check that the configMap has "data.tnf_config.yaml" field, whose value is not empty.

**Secret field:** changed to *string (optional) in the CRD spec. Validation checks:
-   If it's nil,  return.
-   If it's not nil:
    -   Check that it's not an empty string.
    -   Check that the secret exists.
    -   Check that the secret has "data.preflight_dockerconfig.json" field, whose value is not empty.

### CODE REVIEW

File: api/v1alpha1/cnfcertificationsuiterun_webhook.go [hunk 1]
```diff
@@ -97,25 +97,53 @@ func (r *CnfCertificationSuiteRun) ValidateCreate() error {
 
 func (r *CnfCertificationSuiteRun) validateConfigMap() error {
 	configMap := &v1.ConfigMap{}
+
+	// Return an error if config map is not found by name and ns, or field is empty
 	err := c.Get(context.TODO(), types.NamespacedName{Name: r.Spec.ConfigMapName, Namespace: r.Namespace}, configMap)
```
Comment (greyerof) [inline api/v1alpha1/cnfcertificationsuiterun_webhook.go:102]: I think it's a good idea to add a previous explicit check against the `r.Spec.ConfigMapName` not being empty with a custom error message:
```
if r.Spec.ConfigMapName == "" {
  return errors.New("spec.configMapName must not be an empty string")
}
```

File: api/v1alpha1/cnfcertificationsuiterun_webhook.go [hunk 2]
```diff
@@ -97,25 +97,53 @@ func (r *CnfCertificationSuiteRun) ValidateCreate() error {
...
 	if err != nil {
 		logger.Error(err, "CnfCertificationSuiteRun's config map name field is invalid",
```
Comment (greyerof) [inline api/v1alpha1/cnfcertificationsuiterun_webhook.go:104]: Suggestion: we can improve this error msg and the with the configMapName:
```
logger.Error(err, "CnfCertificationSuiteRun's config map name field " + r.Spec.ConfigMapName + " not found in namespace " + r.Namespace)
```

File: api/v1alpha1/cnfcertificationsuiterun_webhook.go [hunk 3]
```diff
@@ -97,25 +97,53 @@ func (r *CnfCertificationSuiteRun) ValidateCreate() error {
...
 			configMapLoggerKey, r.Spec.ConfigMapName)
 		return err
 	}
-	logger.Info("CnfCertificationSuiteRun's config map name field is valid", configMapLoggerKey, configMap.Name)
-	return err
+
+	// Verify required field exists and that it's not empty
+	if value, exists := configMap.Data["tnf_config.yaml"]; !exists || value == "" {
+		err := fmt.Errorf("config map's 'tnf_config.yaml' field must be set")
```
Comment (greyerof) [inline api/v1alpha1/cnfcertificationsuiterun_webhook.go:111]: Please add "... field must be set *with a non-empty and valid configuration yaml for the CNF Certification Suite*"

File: api/v1alpha1/cnfcertificationsuiterun_webhook.go [hunk 4]
```diff
@@ -97,25 +97,53 @@ func (r *CnfCertificationSuiteRun) ValidateCreate() error {
...
+	return nil
 }
 
 func (r *CnfCertificationSuiteRun) validatePreflightSecret() error {
 	preflightSecret := &v1.Secret{}
-	err := c.Get(context.TODO(), types.NamespacedName{Name: r.Spec.PreflightSecretName, Namespace: r.Namespace}, preflightSecret)
+
+	// Nil Preflight Secret is valid
+	if r.Spec.PreflightSecretName == nil {
+		logger.Info("CnfCertificationSuiteRun's preflight secret field is valid", preflightSecretLoggerKey, preflightSecret.Name)
```
Comment (greyerof) [inline api/v1alpha1/cnfcertificationsuiterun_webhook.go:126]: This msg seems a bit misleading to me, as it appears as if the field was set and was valid. You can just inform that no preflight secret name has been set, as a warning:
```
   logger.Info("Warning: CnfCertificationSuiteRun %s (ns %s) has no preflight secret set.", 
   ...
```

File: api/v1alpha1/cnfcertificationsuiterun_webhook.go [hunk 5]
```diff
@@ -97,25 +97,53 @@ func (r *CnfCertificationSuiteRun) ValidateCreate() error {
...
+		return nil
+	}
+
+	// Return an error if preflight secret is not found by name and ns, or field is empty
+	err := c.Get(context.TODO(), types.NamespacedName{Name: *r.Spec.PreflightSecretName, Namespace: r.Namespace}, preflightSecret)
 	if err != nil {
 		logger.Error(err, "CnfCertificationSuiteRun's preflight secret name field is invalid",
```
Comment (greyerof) [inline api/v1alpha1/cnfcertificationsuiterun_webhook.go:133]: Same suggestion as the one I made at line 104.

File: api/v1alpha1/cnfcertificationsuiterun_webhook.go [hunk 6]
```diff
@@ -97,25 +97,53 @@ func (r *CnfCertificationSuiteRun) ValidateCreate() error {
...
 			preflightSecretLoggerKey, r.Spec.PreflightSecretName)
 		return err
 	}
-	logger.Info("CnfCertificationSuiteRun's preflight secret name field is valid", preflightSecretLoggerKey, preflightSecret.Name)
+
+	// Verify required field exists and that it's not empty
+	if value, exists := preflightSecret.Data["preflight_dockerconfig.json"]; !exists || value == nil {
+		err := fmt.Errorf("preflight secret's 'preflight_dockerconfig.json' field must be set")
```
Comment (greyerof) [inline api/v1alpha1/cnfcertificationsuiterun_webhook.go:140]: Similar comment as at line 111. In this case:
"... field must be set *with a valid docker config json content*"

File: api/v1alpha1/cnfcertificationsuiterun_webhook.go (hunk 2 above)
Comment (shirmoran) [reply api/v1alpha1/cnfcertificationsuiterun_webhook.go:104]: The configMapName appears using "key-value" option of the logger. Do you rather add it to the msg it self?

**Review by greyerof — APPROVED** [review]

**Pull request merged**